arch/x86/core/userspace.S

```diff
@@ -8,46 +8,80 @@
 #include <arch/x86/asm.h>
 #include <arch/cpu.h>
 #include <offsets_short.h>
+#include <syscall.h>
 
 /* Exports */
 GTEXT(_x86_syscall_entry_stub)
 GTEXT(_x86_userspace_enter)
 
 /* Imports */
-GTEXT(_k_syscall_entry)
+GTEXT(_k_syscall_table)
 
 /* Landing site for syscall SW IRQ. Marshal arguments and call C function for
- * further processing.
+ * further processing. We're on the kernel stack for the invoking thread.
  */
 SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
-	push	%esi		/* call_id */
+	sti			/* re-enable interrupts */
+
+	/* Marshal arguments per calling convention to match what is expected
+	 * for _k_syscall_handler_t functions
+	 */
+	push	%esp		/* ssf */
 	push	%edi		/* arg5 */
 	push	%ebx		/* arg4 */
 #ifndef CONFIG_X86_IAMCU
+	/* SYS V Convention */
 	push	%ecx		/* arg3 */
 	push	%edx		/* arg2	*/
 	push	%eax		/* arg1 */
 #endif
 
-	call	_k_syscall_entry
+	/* call_id is in ESI. bounds-check it, must be less than
+	 * K_SYSCALL_LIMIT
+	 */
+	cmp	$K_SYSCALL_LIMIT, %esi
```

Comment:
No need to do this now, but if the syscall table is limited to 256 elements, `_k_syscall_table` can be 256-elements long, with all pointers pointing to the oops syscall.  Then use just the lower 8 bits of `esi` to index the table. You'll save  a branch here.

Reply:
Might make this a Kconfig option. On X86 you need tons of RAM anyway due to page alignment but on an MPU-based system could be a toggle.

Reply:
Counterargument: on Lakemont branches aren't nearly as expensive on bigcore (and this jump is forwards, so it's unlikely to be mispredicted in the non-error hot path) and the instruction to mask off the lower 8 bits (ESI has no byte subregisters) is actually going to be longer than the jump, polluting the instruction cache.  Not sure how I'd put money down on which was worse.